src/serac/physics/materials/thermal_functional_material.hpp

```diff
@@ -0,0 +1,278 @@
+// Copyright (c) 2019-2021, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/**
+ * @file thermal_functional_material.hpp
+ *
+ * @brief The material and load types for the thermal functional physics module
+ */
+
+#pragma once
+
+#include "serac/numerics/functional/functional.hpp"
+
+/// ThermalConductionFunctional helper structs
+namespace serac::Thermal {
+
+/// Linear isotropic thermal conduction material model
+class LinearIsotropicConductor {
+public:
+  /**
+   * @brief Construct a new Linear Isotropic Conductor object
+   *
+   * @param density Density of the material (mass/volume)
+   * @param specific_heat_capacity Specific heat capacity of the material (energy / (mass * temp))
+   * @param conductivity Thermal conductivity of the material (power / (length * temp))
+   */
+  LinearIsotropicConductor(double density = 1.0, double specific_heat_capacity = 1.0, double conductivity = 1.0)
+      : density_(density), specific_heat_capacity_(specific_heat_capacity), conductivity_(conductivity)
+  {
+    SLIC_ERROR_ROOT_IF(conductivity_ < 0.0,
+                       "Conductivity must be positive in the linear isotropic conductor material model.");
+
+    SLIC_ERROR_ROOT_IF(density_ < 0.0, "Density must be positive in the linear isotropic conductor material model.");
+
+    SLIC_ERROR_ROOT_IF(specific_heat_capacity_ < 0.0,
+                       "Specific heat capacity must be positive in the linear isotropic conductor material model.");
+  }
+
+  /**
+   * @brief Function defining the thermal flux (constitutive response)
+   *
+   * @tparam T1 type of the temperature (e.g. tensor or dual type)
+   * @tparam T2 type of the temperature gradient (e.g. tensor or dual type)
+   * @param temperature_gradient Gradient of the temperature (du_dx)
+   * @return The thermal flux of the material model
+   */
+  template <typename T1, typename T2>
+  SERAC_HOST_DEVICE T2 operator()(T1& /* temperature */, T2& temperature_gradient) const
+  {
+    return -1.0 * conductivity_ * temperature_gradient;
+  }
```

Comment:
I think these arguments (and others in similar materials/member functions) should be `const T &`. This shouldn't change the behavior, but the current implementation might give the impression that one is allowed to modify them and have `serac::Functional` do something with the mutated values (it doesn't).

The only ones that should be non-`const` references are material state data, since the q-function will need to be able to update that information in e.g. a plasticity model and save the updated values.